### PR TITLE
Add more data form tests

### PR DESCRIPTION
**Pull request by chrzaszcz**
Increase test coverage for [data forms](https://xmpp.org/extensions/xep-0004.html), mainly by testing error handling.

- [x] Please merge the related [Escalus PR](https://github.com/esl/escalus/pull/257) first - I will remove that last commit when a new version of `escalus` is published to `hex.pm`

Other changes:
- Export all functions from test suites and utilities - we are already doing it for other tests, and it makes sense to simplify test development by avoiding lengthy exports that need constant updating.
- Fix minor issues: handle corner cases (e.g. missing form element), remove dead code, simplify form processing
- Introduce `form_helper` in big tests, mainly for form construction and manipulation (to introduce errors for error handling tests).

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/176950) / small_tests / c0d6b5819fe47abdf71493f2dc7ddf4ded3e8eba
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/176950/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/176950/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-05-31_14.13.40/lib.mongooseim.logs/run.2023-05-31_14.13.41/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/176988) / small_tests / 5e7c2c58f94d277541fb96c2fc8f0bc90b616af8
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/176988/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/176988/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-01_08.35.07/lib.mongooseim.logs/run.2023-06-01_08.35.07/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/177050) / small_tests / c4be0a190ca2f068b01975eacfe8de620e1dc97b
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/177050/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/177050/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-02_13.13.34/lib.mongooseim.logs/run.2023-06-02_13.13.35/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_25_arm64](https://circleci.com/gh/esl/MongooseIM/177078) / small_tests / 3ca6910769c518f97fa245df9b107b6f2efe4070
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/177078/small_tests.25.3-arm64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/177078/small_tests.25.3-arm64/small/ct_run.mongooseim@localhost.2023-06-02_14.08.02/lib.mongooseim.logs/run.2023-06-02_14.08.03/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/177105) / small_tests / fa21fd775751f4bc6226d0b81c96d1e47916010f
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/177105/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/177105/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-05_08.11.05/lib.mongooseim.logs/run.2023-06-05_08.11.05/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/177312) / small_tests / ca70e446f9ba7dd3a6ad2b3177eabd94fbde8207
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/177312/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/177312/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-12_13.37.01/lib.mongooseim.logs/run.2023-06-12_13.37.01/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/177366) / small_tests / b17a3271c7102c7b7be48c4574f1e98d2b56a7e7
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/177366/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/177366/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-12_14.26.09/lib.mongooseim.logs/run.2023-06-12_14.26.10/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/177394) / small_tests / 59fdec6d9040b6d31db2835b5c9a551b0ae51cca
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/177394/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/177394/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-12_15.24.11/lib.mongooseim.logs/run.2023-06-12_15.24.12/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/177522) / small_tests / 995bbf157af25f5b7f567a73cc55cc3dbababeb1
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/177522/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/177522/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-13_09.55.34/lib.mongooseim.logs/run.2023-06-13_09.55.35/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/177549) / small_tests / f3b7f0ff67f706ceb097a6f40c5be4eecbcc72c5
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/177549/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/177549/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-13_14.17.42/lib.mongooseim.logs/run.2023-06-13_14.17.43/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/177575) / small_tests / af9ca42c0df94cbad7a61c97a45d1af95f13bb51
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/177575/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/177575/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-14_07.27.00/lib.mongooseim.logs/run.2023-06-14_07.27.00/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/177632) / small_tests / 79d34e25272137d31aabfe96ef62bac4c9ba10c7
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/177632/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/177632/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-14_10.39.56/lib.mongooseim.logs/run.2023-06-14_10.39.56/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/177665) / small_tests / 38b39d8a7d181becdbd961661c7e006df86037ca
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/177665/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/177665/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-14_13.25.06/lib.mongooseim.logs/run.2023-06-14_13.25.07/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/177690) / small_tests / f2b942c170193d75b01de4820c815568379343cb
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/177690/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/177690/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-14_13.49.20/lib.mongooseim.logs/run.2023-06-14_13.49.21/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/177722) / small_tests / 548314c13e39f98212cea1930c234c86eee7f7ee
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/177722/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/177722/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-14_13.58.21/lib.mongooseim.logs/run.2023-06-14_13.58.22/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/177728) / small_tests / 7f1964c584d664730e2b6f9bd8358652e80b4386
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/177728/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/177728/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-14_13.59.11/lib.mongooseim.logs/run.2023-06-14_13.59.11/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/177767) / small_tests / 279b46238dcf91513825c9ae48ebd5d6a691c40b
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/177767/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/177767/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-14_14.16.34/lib.mongooseim.logs/run.2023-06-14_14.16.35/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/177794) / small_tests / 46a45cbec4bba61962ff0d87c046ab5b64c1ac75
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/177794/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/177794/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-14_14.28.20/lib.mongooseim.logs/run.2023-06-14_14.28.21/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/177823) / small_tests / b66f148c2366a7e165f87d3580e5d9ff9691fab7
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/177823/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/177823/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-14_16.20.18/lib.mongooseim.logs/run.2023-06-14_16.20.18/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_25_arm64](https://circleci.com/gh/esl/MongooseIM/177848) / small_tests / 0efa0037bec9130e46417bbcf048c9bf38c11f30
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4031/177848/small_tests.25.3-arm64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4031/177848/small_tests.25.3-arm64/small/ct_run.mongooseim@localhost.2023-06-15_05.56.14/lib.mongooseim.logs/run.2023-06-15_05.56.14/suite.log.html)

**Review by pawlooss1 — COMMENTED** [review]
Great work 👍 I find `form_helper` especially cool - it helps making tests more clear. I left one question and one refactor proposal.

File: big_tests/tests/muc_SUITE.erl [hunk 1]
```diff
@@ -2711,6 +2744,31 @@ user_unregisters_nick_twice(Config) ->
...
     end).
 
+user_cancels_registration(Config) ->
+    escalus:fresh_story(
+      Config, [{alice, 1}],
+      fun(Alice) ->
+              Form = form_helper:form(#{type => <<"cancel">>}),
+              SetIQ = escalus_stanza:iq_set(?NS_INBAND_REGISTER, [Form]),
+              Req = escalus_stanza:to(SetIQ, muc_helper:muc_host()),
+              escalus:send_iq_and_wait_for_result(Alice, Req)
```
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply big_tests/tests/muc_SUITE.erl:2754]: Actually there is an assertion inside: it has to be a result that corresponds to the request (has the same `id` attribute). this particular request is weird, as it is a no-op, but I didn't want to remove its handling from the code, because it is technically valid. It just does nothing, and the result contains nothing.

File: big_tests/tests/form_helper.erl [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+-include_lib("escalus/include/escalus_xmlns.hrl").
+-include_lib("exml/include/exml.hrl").
+
+%% Form creation
+
+form(Spec) ->
+    #xmlel{name = <<"x">>,
+           attrs = [{<<"xmlns">>, ?NS_DATA_FORMS},
+                    {<<"type">>, maps:get(type, Spec, <<"submit">>)}],
+           children = lists:flatmap(fun(Item) -> form_children(Item, Spec) end, [ns, fields])
```
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply big_tests/tests/form_helper.erl:14]: Theoretically a form can be rendered without `ns` or `fields` in the map, and this why I kept the more sophisticated code - and also not to hide similarity to https://github.com/esl/MongooseIM/blob/master/src/mongoose_data_forms.erl#L107

File: big_tests/tests/muc_SUITE.erl (hunk 1 above)
[the remark this answers is not in the document]
Comment (pawlooss1) [reply big_tests/tests/muc_SUITE.erl:2754]: Ok 👍 Thanks for clarifying

File: big_tests/tests/form_helper.erl (hunk 2 above)
[the remark this answers is not in the document]
Comment (pawlooss1) [reply big_tests/tests/form_helper.erl:14]: Makes sense 👍

**Review by pawlooss1 — APPROVED** [review]
